### PR TITLE
 unlock folders - devices list / or paperkey

### PR DESCRIPTION
**Pull request by MarcoPolo**
@keybase/react-hackers 

@keybase/design 

## Preview
screen shot 2016-03-14 at 11 15 40 am

## Zeplin
screen shot 2016-03-14 at 10 58 02 am

### Passing a component that needs more props than connect can offer:

This can happen if Connect returns one set of Props, but the component actually needed another set of Props.

Example:

```js

export type OwnProps = {foo: number}

export type Props = {
  devices: ?Array<Device>,
  phase: UnlockFoldersState.phase,
  foo: number
}

type SneakyProps = {
  devices: ?Array<Device>,
  phase: UnlockFoldersState.phase,
  foo: number,
  lol: string
}

class Foo extends Component<void, SneakyProps, void> {
  render () {
    return <Render/>
  }
}

type Dispatch = TypedDispatch<UnlockFolderActions>
const Connected = typedConnect(
  ({unlockFolders: {devices, phase}}: TypedState, dispatch: Dispatch, ownProps: OwnProps): Props => ({
    devices,
    phase,
    foo: ownProps.foo
  })
)(Foo)
```

Error:

```
unlock-folders/index.js:28
 28: class Foo extends Component<void, SneakyProps, void> {
                                           ^^^^^^^^^^^ property `lol`. Property not found in
 36:   ({unlockFolders: {devices, phase}}: TypedState, dispatch: Dispatch, ownProps: OwnProps): Props => ({
                                                                                                ^^^^^ object type
```

This tells us that the props that Foo needs isn't being fulfilled by this connect call.

### Not passing in the proper ownProps

Example:

```js
export type OwnProps = {foo: number}

export type Props = {
  devices: ?Array<Device>,
  phase: UnlockFoldersState.phase,
  foo: number
}

class UnlockFolders extends Component<void, Props, void> {
  render () {
    return <Render/>
  }
}

type Dispatch = TypedDispatch<UnlockFolderActions>
const Connected = typedConnect(
  ({unlockFolders: {devices, phase}}: TypedState, dispatch: Dispatch, ownProps: OwnProps): Props => ({
    devices,
    phase,
    foo: ownProps.foo
  })
)(UnlockFolders)

const t = <Connected foo={'NaN!'}/>
```

Error:

```
unlock-folders/index.js:36
 36: const t = <Connected foo={'NaN!'}/>
               ^^^^^^^^^^^^^^^^^^^^^^^^^ React element `Connected`
 36: const t = <Connected foo={'NaN!'}/>
                               ^^^^^^ string. This type is incompatible with
 13: export type OwnProps = {foo: number}
                                  ^^^^^^ number
```

This means that for this component we have typed:

* Reducer
  * Its state
  * Its reducing function
  * actions it operations on
* Actions
  * What they return and what other actions they will dispatch
* Smart Components
  * Know what we will get from a `connect` call (error if we miss or get something wrong)
  * Declare what types you want from `connect` vs what types need to be passed in to the connected component

* Dumb Components
  * Make sure we get all the right props

## Still TODO

Is it possible to type bindActionCreators?

### CODE REVIEW

**Comment by cecileboucheron** [comment]
Looking good! The icons look blurry though, are you using the @2x ones?

File: shared/unlock-folders/device-list.desktop.js [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+  }
+}
+
+export default class DeviceList extends Component<void, Props, void> {
+  render () {
+    return (
+      <div style={{...globalStyles.flexBoxColumn, alignItems: 'center'}}>
+        <Text type='Body' style={styles.infoText}>Turn on one of your devices to unlock your folders:</Text>
+        <div style={styles.devicesContainer}>
+          {this.props.devices && this.props.devices.map(d => <DeviceRow device={d}/>)}
```
Comment (chrisnojima) [inline shared/unlock-folders/device-list.desktop.js]: do you get a key warning here?

File: shared/unlock-folders/device-list.desktop.js [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+export type Props = {
+  devices: ?Array<Device>,
+  toPaperKeyInput: () => void
+}
+
+class DeviceRow extends Component<void, {device: Device}, void> {
+  render () {
+    const isDesktop = this.props.device.type === 'desktop'
+    return (
+      <div style={{...globalStyles.flexBoxRow, marginBottom: 16}}>
```
Comment (chrisnojima) [inline shared/unlock-folders/device-list.desktop.js]: maybe move these styles into the const below

File: shared/unlock-folders/device-list.desktop.js [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+        <div style={{width: 24, marginLeft: 33, display: 'flex', justifyContent: 'center'}}>
+          <Icon type={isDesktop ? 'computer-big' : 'phone-big'} style={{height: 21}}/>
+        </div>
+        <Text type='BodySemibold' style={{...globalStyles.italic, marginLeft: 16, cursor: 'default'}}>{this.props.device.name}</Text>
```
Comment (chrisnojima) [inline shared/unlock-folders/device-list.desktop.js]: This should be put into Text I think, it'll be used elsewhere

File: shared/unlock-folders/device-list.desktop.js [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+    alignSelf: 'center',
+    paddingTop: 15,
+    paddingBottom: 15
+  },
+
+  buttonsContainer: {
+    ...globalStyles.flexBoxRow,
+    marginTop: 15,
+    marginRight: 30,
+    marginLeft: 'auto'
```
Comment (chrisnojima) [inline shared/unlock-folders/device-list.desktop.js]: we don't need this, just do alignSelf: 'flex-end'

File: shared/unlock-folders/device-list.desktop.js [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+  },
+
+  enterPaperKey: {
+    height: 32,
+    width: 236,
+    marginRight: 7
+  },
+
+  accessFolders: {
+    marginRight: 0
```
Comment (chrisnojima) [inline shared/unlock-folders/device-list.desktop.js]: is this needed?

**Comment by chrisnojima** [comment]
small comments :+1:

File: shared/unlock-folders/device-list.desktop.js (hunk 3 above)
Comment (MarcoPolo) [inline shared/unlock-folders/device-list.desktop.js]: You mean italic?

File: shared/unlock-folders/device-list.desktop.js (hunk 4 above)
Comment (MarcoPolo) [inline shared/unlock-folders/device-list.desktop.js]: It's the same thing, will change though

File: shared/unlock-folders/device-list.desktop.js (hunk 5 above)
Comment (MarcoPolo) [inline shared/unlock-folders/device-list.desktop.js]: yes, the default marginRight is 10

File: shared/unlock-folders/device-list.desktop.js (hunk 3 above)
Comment (chrisnojima) [inline shared/unlock-folders/device-list.desktop.js]: this specific italic style, i don't think we'll italicize any other types, i think it'll be just this right @cecileboucheron

**Comment by cecileboucheron** [comment]
@chrisnojima Yeah italic will just be for my own devices, my own username and my own paper keys names throughout the app.

**Pull request merged**